### PR TITLE
Functions for computing a cycle basis, part 2

### PR DESCRIPTION
**Pull request by jfinkels**
Unfortunately, I had some problem with git not uploading my new commits for the pull request #1538; I couldn't figure out how to fix it so I opened this new pull request instead. Sorry about that.

Anyway, the new commit I made capitalizes on the changes introduced in pull request #1740 (the `Graph.edge_subgraph()` method) and in pull request #1741 (allowing keys to be returned by the `minimum_spanning_edges()` function). In fact, I extended the latter pull request in commit 1d4e7dd67115a660aad621845828f9483a2d8373, the first commit in the current pull request, since I needed access to the edge key for the edges from which the minimum spanning tree of a multigraph was derived. The `chords()` function is now as simple as I had originally hoped.

### CODE REVIEW

File: examples/algorithms/kirchhoffs_circuit_laws.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python
```
Comment (dschult) [inline examples/algorithms/kirchhoffs_circuit_laws.py]: Should kirchoffs_circuit_laws.py go with the examples?

Comment (dschult) [inline examples/algorithms/kirchhoffs_circuit_laws.py]: oops -- I see that this file is in the examples directory already...

Should we remove the ```if __name__ == "__main__":```?

Comment (jfinkels) [inline examples/algorithms/kirchhoffs_circuit_laws.py]: Sure, I'll update this file.

**Comment by jfinkels** [comment]
There seems to be problems with the unit tests, some are not actually being run. I need to do a bit more reorganizing on this pull request.

**Comment by jfinkels** [comment]
Due to duplicate test method names, the unit tests for the proposed `cycle_basis_matrix` function  were not being run. I have fixed the naming issues, and the tests now fail.

Maybe it would be easier to deal with if I split this quite large pull request into several commits, each of which depends on the one above it:

1. adding the `as_multigraph` keyword argument to the `minimum_spanning_tree` function
2. adding the `chords` function
3. adding the `cycle_basis_matrix` function
4. adding the Kirchhoff circuit laws example

What do you think?

**Comment by dschult** [comment]
I think the tests are failing because "an arbitrary orientation of the edges is selected".  Perhaps the test could check the absolute value of the matrix?

So, I don't think you need to rearrange the commits unless it helps you with debugging.

Is there a way to know (or maybe even determine) the order of the edges and cycles in the matrix?
That questions makes me realize that this feature has similarities with the laplacian and adjacency_matrix code.  Should cycle_basis_matrix be a module inside the linalg directory?
I'm fine with leaving it in cycles.py but fine with moving it too.

**Comment by dschult** [comment]
Why doesn't the ```cycle_basis_matrix``` function use the existing ```cycle_basis``` function and ```utils.misc.pairwise(_, cyclic=True)``` routines iterating over cycles and adding entries to the matrix?
Is this algorithm different? better? cited anywhere?

**Comment by kakila** [comment]
Dear Developers,
I am one of the original authors of the pull request and I was wondering if this was finally included or not in the software. As far as I can tell, the conversion to cycle matrix and the examples are not released, is this correct?
Please let me know if I can assist.

Regards,

**Comment by kakila** [comment]
I think the change to add the optional argument to the spanning trees (commit 1d4e7dd) could be integrated right away. This is a useful enhancement independent of the cycle basis topic

**Comment by dschult** [comment]
This should be replaced by #4822 
Pick up the conversation on that thread.
Leaving this open for now in case the rebase of that one is too confusing.

**Pull request closed without merging**